protocol-units/bridge/chains/ethereum/src/event_monitoring.rs

```diff
@@ -0,0 +1,482 @@
+use crate::types::{
+	EthAddress, EthHash, EventName, COUNTERPARTY_ABORTED_SELECT, COUNTERPARTY_COMPLETED_SELECT,
+	COUNTERPARTY_LOCKED_SELECT, INITIATOR_COMPLETED_SELECT, INITIATOR_INITIATED_SELECT,
+	INITIATOR_REFUNDED_SELECT,
+};
+use crate::EthChainEvent;
+use alloy::dyn_abi::EventExt;
+use alloy::eips::BlockNumberOrTag;
+use alloy::primitives::{address, LogData};
+use alloy::providers::{Provider, ProviderBuilder, RootProvider, WsConnect};
+use alloy::rpc::types::{Filter, Log};
+use alloy::{
+	json_abi::{Event, EventParam},
+	pubsub::PubSubFrontend,
+};
+use bridge_shared::bridge_monitoring::{
+	BridgeContractCounterpartyEvent, BridgeContractCounterpartyMonitoring,
+};
+use bridge_shared::types::{Amount, LockDetails, SmartContractInitiatorEvent, TimeLock};
+use bridge_shared::{
+	bridge_monitoring::{BridgeContractInitiatorEvent, BridgeContractInitiatorMonitoring},
+	types::{
+		BridgeTransferDetails, BridgeTransferId, HashLock, InitiatorAddress, RecipientAddress,
+		SmartContractCounterpartyEvent,
+	},
+};
+use futures::{channel::mpsc::UnboundedReceiver, Stream, StreamExt};
+use std::{pin::Pin, task::Poll};
+
+pub struct EthInitiatorMonitoring<A, H> {
+	listener: UnboundedReceiver<EthChainEvent<A, H>>,
+	_ws: RootProvider<PubSubFrontend>,
+}
+
+impl BridgeContractInitiatorMonitoring for EthInitiatorMonitoring<EthAddress, EthHash> {
+	type Address = EthAddress;
+	type Hash = EthHash;
+}
+
+impl EthInitiatorMonitoring<EthAddress, EthHash> {
+	pub async fn build(
+		rpc_url: &str,
+		listener: UnboundedReceiver<EthChainEvent<EthAddress, EthHash>>,
+	) -> Result<Self, anyhow::Error> {
+		let ws = WsConnect::new(rpc_url);
+		let ws = ProviderBuilder::new().on_ws(ws).await?;
+
+		//TODO: this should be an arg
+		let initiator_address = address!("f39Fd6e51aad88F6F4ce6aB8827279cffFb92266");
+		let filter = Filter::new()
+			.address(initiator_address)
+			.event("BridgeTransferInitiated(bytes32,address,bytes32,uint256)")
+			.event("BridgeTransferCompleted(bytes32,bytes32)")
+			.from_block(BlockNumberOrTag::Latest);
+
+		let sub = ws.subscribe_logs(&filter).await?;
+		let mut sub_stream = sub.into_stream();
+
+		// Spawn a task to forward events to the listener channel
+		let (sender, _) =
+			tokio::sync::mpsc::unbounded_channel::<EthChainEvent<EthAddress, EthHash>>();
```

Comment:
If the listener is dropped, sent message won't be read. There's an issue with the listener in parameter and this send. They aren't connected, but the code supposes as they are. I've corrected in the PR #624.

Reply:
Yes a noticed this, as separate listeners, if you've corrected in 624, this can go in like this and then 624 will follow up when ready.